File: micro/fn_call.c
// C equivalent of fn_call.hl. Free-fn call overhead, minimal body.
// __attribute__((noinline)) forces a real call so it measures call
// overhead, not a folded-away loop. Caveat: unlike Go's
// //go:noinline, clang still does IPA purity analysis on noinline
// fns and would delete the dead pure calls — the asm barrier marks
// the result observed each iteration to keep every call real.
#include <stdio.h>
#include <time.h>

static long long now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (long long)ts.tv_sec * 1000000000LL + (long long)ts.tv_nsec;
}

__attribute__((noinline)) static long long step(long long x) { return x * 2 + 1; }

int main(void) {
    long long iters = 10000000;
    long long t0 = now_ns();
    long long acc = 0;
    for (long long i = 0; i < iters; i++) {
        acc = step(i);
        asm volatile("" :: "r"(acc));
    }
    long long elapsed = now_ns() - t0;
    printf("iters=%lld\n", iters);
    printf("acc=%lld\n", acc);
    printf("elapsed_ns=%lld\n", elapsed);
    return 0;
}

File: micro/fn_call.rs
// Rust equivalent of fn_call.hl. Free-fn call overhead, minimal body.
// #[inline(never)] forces a real call; unlike //go:noinline, LLVM still
// DCEs dead calls to fns it infers as pure, so each result is pinned
// with black_box to keep the call live every iteration.

use std::hint::black_box;
use std::time::Instant;

#[inline(never)]
fn step(x: i64) -> i64 {
    x * 2 + 1
}

fn main() {
    let iters: i64 = 10000000;
    let t0 = Instant::now();
    let mut acc: i64 = 0;
    for i in 0..iters {
        acc = black_box(step(i));
    }
    let elapsed = t0.elapsed().as_nanos();
    println!("iters={}", iters);
    println!("acc={}", acc);
    println!("elapsed_ns={}", elapsed);
}
